packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts

```diff
@@ -327,12 +364,121 @@ export class AutomergeHost extends Resource {
     const handle = this._repo.handles[documentId];
     if (handle) {
       const doc = handle.docSync();
-      if (!doc) {
-        return undefined;
+      if (doc) {
+        return getHeads(doc);
+      }
+    }
+    return this._headsStore.getHeads(documentId);
+  }
+
+  //
+  // Collection sync.
+  //
+
+  getLocalCollectionState(collectionId: string): CollectionState | undefined {
+    return this._collectionSynchronizer.getLocalCollectionState(collectionId);
+  }
+
+  getRemoteCollectionStates(collectionId: string): ReadonlyMap<PeerId, CollectionState> {
+    return this._collectionSynchronizer.getRemoteCollectionStates(collectionId);
+  }
+
+  refreshCollection(collectionId: string) {
+    this._collectionSynchronizer.refreshCollection(collectionId);
+  }
+
+  async getCollectionSyncState(collectionId: string): Promise<CollectionSyncState> {
+    const result: CollectionSyncState = {
+      peers: [],
+    };
+
+    const localState = this.getLocalCollectionState(collectionId);
+    const remoteState = this.getRemoteCollectionStates(collectionId);
+
+    if (!localState) {
+      return result;
+    }
+
+    for (const [peerId, state] of remoteState) {
+      const diff = diffCollectionState(localState, state);
+      result.peers.push({
+        peerId,
+        differentDocuments: diff.different.length,
+      });
+    }
+
+    return result;
+  }
+
+  /**
+   * Update the local collection state based on the locally stored document heads.
+   */
+  async updateLocalCollectionState(collectionId: string, documentIds: DocumentId[]) {
+    const heads: Record<DocumentId, Heads> = {};
+    await Promise.all(
+      documentIds.map(async (documentId) => {
```

Comment:
not sure we want to crash the whole thing if we fail to retrieve heads for a document. we'll retry eventually, so I'd just log and swallow the error here.

also I think it's better to use getMany queries in such cases, even remembering how "nice" the implementation is in browsers. if the implementation changes we'll benefit from it. also they have parallel limit there which might be good for spreading the load in time if documentIds collection is big.

Reply:
👍

Reply:
getHeads does only local IO, im not sure if we need to gracefully handle storage failures

Reply:
TODO: test how getMany handles missing keys -- we want to return undefined for missing ones, and not fail the whole query